base-files: upgrade: kill remaining processes only once

mcprat opened the pull request:
Some processes are managed by procd and set to restart when killed
via the procd instance parameter "respawn" being set during init.

Example:
procd_set_param respawn 3600 1 0

The function kill_remaining() is called twice,
and generates a new list of processes each time.
This causes some of the same procd managed daemons
to be killed and restarted twice instead of once.

Therefore, let the list of processes be generated in a separate function
and let kill_remaining() use that same list each time it is called.
This way, the list of processes to be killed is static and PID-specific,
so when a managed daemon restarts, it will not be killed again
due to having a different PID.

Also, sleep for a few seconds before the list is generated
so that processes already terminated are not included
in the list to be killed again.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
base-files: upgrade: kill remaining processes only once

Some processes are managed by procd and set to restart when killed
via the procd instance parameter "respawn" being set during init.

Example:
procd_set_param respawn 3600 1 0

The function kill_remaining() is called twice,
and generates a new list of processes each time.
This causes some of the same procd managed daemons
to be killed and restarted twice instead of once.

Therefore, let the list of processes be generated in a separate function
and let kill_remaining() use that same list each time it is called.
This way, the list of processes to be killed is static and PID-specific,
so when a managed daemon restarts, it will not be killed again
due to having a different PID.

Also, sleep for a few seconds before the list is generated
so that processes already terminated are not included
in the list to be killed again.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mpratt14 commented:
@aparcar another small fix I discovered a few days ago

wulfy23 commented on package/base-files/files/lib/upgrade/stage2:
```diff
@@ -76,40 +76,56 @@ switch_to_ramfs() {
+		name="${name#(}"; name="${name%)}"
+
+		# Skip PID1, our parent, ourself and our children
+		[ $pid -ne 1 -a $pid -ne $proc_ppid -a $pid -ne $$ -a $ppid -ne $$ ] || continue
```
```
[ "$ppid" = "I"  ] && continue
```

mpratt14 replied:
sorry I don't get it, is that a typo?

wulfy23 replied:
sh out of range?

mpratt14 replied:
`$ppid` is a number and `I` is a letter....

wulfy23 replied:
exactly

mpratt14 replied:
note that I copied this line from below

this conditional needs to pass for some processes and fail for others, if I match ppid to a letter, it will either always pass or always fail....

adschm replied:
Maybe you assign some variables differently...? Apart from that, I'd not touch this copied code unless necessary due to the copying itself.

mpratt14 replied:
I see nothing wrong with the line as it is, it works

wulfy23 replied:
may work for you... but not for me... your pr... your call

i've patched this locally as i'm not a fan of errors during
the sysupgrade process....
```
#!/bin/sh

for stat in /proc/[0-9]*/stat; do
	[ -f "$stat" ] || continue
	read pid name state ppid rest < $stat 2>/dev/null    
        name="${name#(}"; name="${name%)}"
	echo "ppid:$ppid name:$name"
done

#######################################
[snip]
ppid:1 name:logd
ppid:1 name:rpcd
ppid:1 name:dropbear
ppid:1 name:netifd
ppid:1 name:odhcpd
ppid:1 name:crond
ppid:1 name:ubusd
ppid:4582 name:udhcpc
ppid:4582 name:odhcp6c
ppid:2 name:kblockd
ppid:1 name:askfirst
ppid:1 name:rngd
ppid:2 name:irq/39-aerdrv
ppid:2 name:hwrng
ppid:2 name:iscsi_eh
ppid:2 name:kworker/3:2H
ppid:I name:DWC
ppid:2 name:uas
ppid:2 name:kworker/u8:0-events_unbound
ppid:2 name:vchiq-slot/0
ppid:2 name:ipv6_addrconf
[snip]
##########################################
```

mpratt14 replied:
what's not working for you?

also, the read command will not error because the previous line checks that the input file exists, and otherwise skips it

wulfy23 replied:
some kernel process contain spaces...

while the read command does not error
the numerical comparison as mentioned
does with the phrase "sh out of range"
when it encounters the "I" char sucked
in by read and its failure to handle spaces
in "value 2" from stat

as we won't be killing kernel threads
anyway... I see no need to implement
more complex workarounds

continuing on the "I" char prior to the
numerical comparisons handle this
case well

mpratt14 replied:
ok I understand now

do you have an example of a process with a name, so I can test?

mpratt14 replied:
I think the solution is to not strip the parenthesis from the stat file, rather replace with quotes?

mpratt14 replied:
the space would still cause read to split it...hmmm

wulfy23 replied:
its and edge case... but the dd redirect shennanigans do provide
precendent for minimal false errors and crap during sysupgrade...

i'm not sure of the offender in my case "DWC Notifcation" is even
compliant with standards...

would be interesting to post the sample above on the forum and see
just how many people / procs match this...

for clean handling... (if seen as desired) 'shift on alpha' may work
fairly simplistically...

mpratt14 replied:
@wulfy23 see the new commit and try it out 👍🏼

wulfy23 replied:
damn... this implementation for handling
spaces in stat name has passed the
wulfy23 nitpick stresstest

```
ppid:2 name:DWC_Notificatio
```
;)

aparcar replied:
Uhm who would have a whitespace in the process name? I don't think this is really required sorry...

mpratt14 replied:
It would be best to correct that one program to not have a space in its name...

@wulfy23 where did it come from?

wulfy23 replied:
the kernel: drivers/usb/host/dwc_common_port/dwc_notifier.c

mcprat committed:
base-files: upgrade: handle processes with whitespace in the name

There are rare situations where a process name has a space.
In this case, read command makes PPID equal to process State or something else...
Then, an error goes to stdout and the process is not killed with the others.

It's not possible to pipe a string variable into the read command,
so use shell expansions to define the pieces from the stat file.

Reported-by: Imran Khan <gururug@gmail.com>
Signed-off-by: Michael Pratt <mcpratt@pm.me>

mpratt14 commented:
OK, now it is 2 small fixes, and ready 👍🏼

aparcar commented:
Sorry could you please explain this again? Didn't we just merge something to avoid respawning?

NeoRaider commented:
I don't get why this change is necessary.

- procd is not running anymore when stage2 is executed (it has replaced itself with *upgraded* as PID1), so no services are restarted anyways
- We really must try to kill everything as hard as possible. This doesn't need to be efficient, but it *must* catch every single process that is running from the original rootfs and not the ramdisk, or unmounting the rootfs will fail. This is relevant for all targets that don't have redundant rootfs, so the system that is running must overwrite itself.

mpratt14 commented:
> * procd is not running anymore when stage2 is executed (it has replaced itself with _upgraded_ as PID1),

ah....I was not aware of this....

then I was wasting time after all...

the second commit I just added is still useful though....

mpratt14 commented:
sorry to waste everyone's time on this

@wulfy23 im going to send a patch to raspi kernel to fix that process name

wulfy23 commented:
```
Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Jonathan Bell <jonathan@raspberrypi.org>
Signed-off-by: Kari Suvanto <karis79@gmail.com>
Signed-off-by: Lukas Wunner <lukas@wunner.de>
Signed-off-by: Malik Olivier Boussejra <malik@boussejra.com>
Signed-off-by: Marek Behún <marek.behun@nic.cz>
Signed-off-by: Michael Zoran <mzoran@crowfest.net>
Signed-off-by: Nathan Chancellor <natechancellor@gmail.com>
Signed-off-by: Noralf Trønnes <noralf@tronnes.org>
Signed-off-by: Paul Zimmerman <paulz@synopsys.com>
Signed-off-by: Phil Elwell <phil@raspberrypi.org>
Signed-off-by: popcornmix <popcornmix@gmail.com>
```

it's within spec... and stage2 will still have this bug...

```
### cat /proc/666/stat
666 (brick(me)) I 101

aparcar commented:
> sorry to waste everyone's time on this

No worries, happens to all of us. Please don't feel discouraged to do further contributions!

wulfy23 commented:
> stage2 will still have this bug

fixed today by dango 4d3303b1bddf3c66fc0a05fb5eb45ceac48cd0d5

thankyou!
